### ISSUE TITLE
verify and document up_frac_i and up_frac_f calculations for magnetic models (Trac #1086)

### ISSUE THREAD

**Issue by pkienzle** [SasView/sasview#1137]
The weights for each cross section computed from up_frac_i and up_frac_f need to be checked.  The code uses:

```text
    wuu = sqrt(up_i * up_f)
    wud = sqrt(up_i * (1-up_f))
    wdu = sqrt((1-up_i) * up_f)
    wdd = sqrt((1-up_i) * (1-up_f))
    I = wuu*Iuu + wud*Iud + wdu*Idu + wdd*Idd 
```

If either up_i or up_f is not 0 or 1, then the sum of the weights does not equal one which does not sound correct.  That is, if up_i/up_f represents leakage between channels, the total number of neutrons should be preserved.

Consider up_i = 0.9, which lets through 90% up and 10% down from the guides (so ```up/(up+down)=0.9```, and similarly up_f=0.9, which lets through 90% up and 10% down after the sample.  Tracing the individual channels:

```text
initial =>  sample   =>  final selection

 up=0.9 =>  0.9*Iuu  =>  0.9*0.9*Iuu = 0.81*Iuu   [true pos]
                         0.1*0.9*Iuu              [false neg]
            0.9*Iud  =>  0.1*0.9*Iud = 0.09*Iud   [false pos]
                         0.9*0.9*Iud              [true neg] 
 dn=0.1 =>  0.1*Idu  =>  0.9*0.1*Idu = 0.09*Idu   [true pos]
                         0.1*0.1*Idu              [false neg]
            0.1*Idd  =>  0.1*0.1*Idd = 0.01*Idd   [false pos]
                         0.9*0.1*Idd              [true neg]
```
**So weights should not use sqrt**.  Note that the false positive and false negative rates may be independent, and further, with He3 polarizers, may be time dependent, but we keep things simple by having only one number per polarizer rather than two.

Doing the same analysis with a partially polarized or unpolarized beam requires an additional factor of two intensity correction.  In this case, with 90% efficiency on the front end and 50% on the back end:

```text
initial  => sample  => final selection
 up=0.9 =>  0.9*Iuu  =>  0.5*0.9*Iuu = 0.45*Iuu
                         0.5*0.9*Iuu
            0.9*Iud  =>  0.5*0.9*Iud = 0.45*Iud
                         0.5*0.9*Iud
 dn=0.1 =>  0.1*Idu  =>  0.5*0.1*Idu = 0.05*Idu
                         0.5*0.1*Idu
            0.1*Idd  =>  0.5*0.1*Idd = 0.05*Idd
                         0.5*0.1*Idd
```

For non-magnetic systems ```Iuu=Idd``` and ```Iud=Idu=0```.  In the first example, a correction of ```1/(up_i*up*f + (1-up_i)*(1-up_f))``` is needed so that ```I=Iuu=Iud```.  In the second example the correction is ```2```.

Update the user guide to contain complete details of the final implementation and how it is to be interpreted.

Migrated from http://trac.sasview.org/ticket/1086
```json
{
    "status": "reopened",
    "changetime": "2019-03-29T10:14:54",
    "_ts": "2019-03-29 10:14:54.453013+00:00",
    "description": "The weights for each cross section computed from up_frac_i and up_frac_f need to be checked.  The code uses:\n{{{\n    wuu = sqrt(up_i * up_f)\n    wud = sqrt(up_i * (1-up_f))\n    wdu = sqrt((1-up_i) * up_f)\n    wdd = sqrt((1-up_i) * (1-up_f))\n    I = wuu*Iuu + wud*Iud + wdu*Idu + wdd*Idd \n}}}\n\nIf either up_i or up_f is not 0 or 1, then the sum of the weights does not equal one which does not sound correct.  That is, if up_i/up_f represents leakage between channels, the total number of neutrons should be preserved.\n\nConsider up_i = 0.9, which lets through 90% up and 10% down from the guides (so {{{up/(up+down)=0.9}}}, and similarly up_f=0.9, which lets through 90% up and 10% down after the sample.  Tracing the individual channels:\n{{{\ninitial =>  sample   =>  final selection\n\n up=0.9 =>  0.9*Iuu  =>  0.9*0.9*Iuu = 0.81*Iuu   [true pos]\n                         0.1*0.9*Iuu              [false neg]\n            0.9*Iud  =>  0.1*0.9*Iud = 0.09*Iud   [false pos]\n                         0.9*0.9*Iud              [true neg] \n dn=0.1 =>  0.1*Idu  =>  0.9*0.1*Idu = 0.09*Idu   [true pos]\n                         0.1*0.1*Idu              [false neg]\n            0.1*Idd  =>  0.1*0.1*Idd = 0.01*Idd   [false pos]\n                         0.9*0.1*Idd              [true neg]\n}}}\n**So weights should not use sqrt**.  Note that the false positive and false negative rates may be independent, and further, with He3 polarizers, may be time dependent, but we keep things simple by having only one number per polarizer rather than two.\n\nDoing the same analysis with a partially polarized or unpolarized beam requires an additional factor of two intensity correction.  In this case, with 90% efficiency on the front end and 50% on the back end:\n{{{\ninitial  => sample  => final selection\n up=0.9 =>  0.9*Iuu  =>  0.5*0.9*Iuu = 0.45*Iuu\n                         0.5*0.9*Iuu\n            0.9*Iud  =>  0.5*0.9*Iud = 0.45*Iud\n                         0.5*0.9*Iud\n dn=0.1 =>  0.1*Idu  =>  0.5*0.1*Idu = 0.05*Idu\n                         0.5*0.1*Idu\n            0.1*Idd  =>  0.5*0.1*Idd = 0.05*Idd\n                         0.5*0.1*Idd\n}}}\n\nFor non-magnetic systems {{{Iuu=Idd}}} and {{{Iud=Idu=0}}}.  In the first example, a correction of {{{1/(up_i*up*f + (1-up_i)*(1-up_f))}}} is needed so that {{{I=Iuu=Iud}}}.  In the second example the correction is {{{2}}}.\n\nUpdate the user guide to contain complete details of the final implementation and how it is to be interpreted.",
    "reporter": "pkienzle",
    "cc": "",
    "resolution": "",
    "workpackage": "SasView Bug Fixing",
    "time": "2018-04-06T22:18:01",
    "component": "SasView",
    "summary": "verify and document up_frac_i and up_frac_f calculations for magnetic models",
    "priority": "major",
    "keywords": "",
    "milestone": "SasView 4.3.0",
    "owner": "",
    "type": "defect"
}
```

**Comment by pkienzle** [comment]
Trac update at `2018/04/07 14:40:00`: **pkienzle** commented:

This code exists in ```sascalc/calculator/c_extensions/libfunc.c:cal_msld``` and ```sasmodels/kernel_iq.c:set_spin_weights```

**Comment by pkienzle** [comment]
Trac update at `2018/04/09 14:36:36`: **pkienzle** commented:

This analysis works regardless of whether selectivity is based on spin leakage in the polarizer/analyser or whether it is due to depolarization in the guide field.

On the front end, leakage of down through the front polarizer can give the same ```up/(up+down)``` ratio entering the sample as a perfect up selection in the polarizer followed by percentage depolarization.  On the back end with non-spin-flip (NSF) samples, false positive down neutrons from ```Idd``` are indistinguishable from a portion of the down neutrons from ```Idd``` flipping to up and being accepted by a perfect analyser.  For spin-flip (SF) samples with ```Iud==Idu``` (the usual case) depolarization is again indistinguishable from analyser selectivity.  Only case that may cause problems is SF with ```Iud!=Idu``` since then it matters whether the sample sees up vs down neutrons, but this never occurs with the simple model of magnetism we are using (the imaginary sld is ```+mz``` for ```Iud``` and ```-mz``` for ```Idu``` so the squared contrast is always the same).  Such samples will need specialized 2D models which allow magnetism that varies within the sample, and these can define their own notion of spin up fraction if they need it.

**Comment by rprospero** [comment]
Trac update at `2018/04/17 20:45:00`: **awashington** commented:

I spoke with Rob Dalgliesh about this earlier.  We agree that we should be calculating intensities and not amplitudes, so we should not be taking the square roots.  The square roots might make sense if we were calculating an amplitude and there was going to be interference between the spin states but

1. The initial unpolarised beam is a mixed state that is operated on by the polariser and analyser.  These mixed states will not interfere with each other.  Sasview 9 might consider allowing for misaligned flippers that do not perform a full pi rotation, but that is well outside of our current scope.
2. The weight is being multiplied onto the scattered intensity, not the scattered amplitude, so the treating the weights as amplitudes wouldn't make any sense.

**Comment by rprospero** [comment]
Trac update at `2018/06/04 08:14:58`: **awashington** commented:

Fixed for sasmodels in pr 69.

I haven't touched this yet for sasview, because genicomXY is different from the primary scattering code and looks to be calculating an actual amplitude.  I'm basing this mostly on the fact that it's calculating scattering from individual nuclei, so I'm guessing that we're on a length scale where interference will be important, so the amplitude will be necessary.  Plus the fact that everything si complex and the amplitude is calculated at the end.

In this case, I believe that we do need to keep the square root terms, since the amplitude is squared at the end of the function.  I think that we would need a sqrt term, but not perhaps the fourth root that we currently have?

**Comment by rprospero** [comment]
Trac update at `2018/07/04 13:30:51`: **awashington** commented:

The use of spin fraction as the base value is certainly muddying the waters, but I think that I may still be misunderstanding something.  I would have argued that the weights cannot be the polarisation because that would allow for negative weights when there is negative polarisation, which I would have thought to be non-physical.

**Comment by butlerpd** [comment]
Trac update at `2018/07/24 14:03:31`: 
* **butler** commented:

We probably should have a separate conf call to sort out this issue?  At the moment it doesn't seem like it can be resolved for 4.2 so going to move it to 4.3 at this point.  Should perhaps be noted in release notes for 4.2?


* **butler** changed milestone from "SasView 4.2.0" to "SasView 4.3.0"

**Comment by dehoni** [comment]
This issue is fixed since March 2019. The weighting with incoming and outgoing polarised neutron allow to produce scattering cross section and any combination thereof consistently.
